Solaris 11, miniupnpc patch

When compiling ***miniupnpc*** on Solaris 11:

```
gcc -m64 -fPIC -pipe -O1 -I/export/home/zcash/depends/x86_64-unknown-linux-gnu/include -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600   -c -o igd_desc_parse.o igd_desc_parse.c
gcc -m64 -fPIC -pipe -O1 -I/export/home/zcash/depends/x86_64-unknown-linux-gnu/include -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600   -c -o minisoap.o minisoap.c
In file included from /usr/include/iso/string_iso.h:24:0,
                 from /usr/include/string.h:11,
                 from minixml.c:33:
/usr/include/sys/feature_tests.h:358:2: error: #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications 	require the use of c99"
 #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications \
  ^
make[1]: *** [minixml.o] Error 1
make[1]: *** Waiting for unfinished jobs....
In file included from /usr/include/stdio.h:15:0,
                 from minisoap.c:10:
/usr/include/sys/feature_tests.h:358:2: error: #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications 	require the use of c99"
 #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications \
  ^
In file included from /usr/include/stdio.h:15:0,
                 from miniwget.c:9:
/usr/include/sys/feature_tests.h:358:2: error: #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications 	require the use of c99"
 #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications \
  ^
In file included from /usr/include/stdio.h:15:0,
                 from igd_desc_parse.c:10:
/usr/include/sys/feature_tests.h:358:2: error: #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications 	require the use of c99"
 #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications \
  ^
make[1]: *** [igd_desc_parse.o] Error 1
make[1]: *** [minisoap.o] Error 1
miniwget.c: In function ‘getHTTPResponse’:
miniwget.c:87:6: warning: variable ‘reason_phrase_len’ set but not used [-Wunused-but-set-variable]
  int reason_phrase_len = 0;
      ^
miniwget.c:86:9: warning: variable ‘reason_phrase’ set but not used [-Wunused-but-set-variable]
  char * reason_phrase = NULL;
         ^
make[1]: *** [miniwget.o] Error 1
make[1]: Leaving directory `/export/home/zcash/depends/work/build/x86_64-unknown-linux-gnu/miniupnpc/2.0-792e2b4396f'
make: *** [/export/home/zcash/depends/work/build/x86_64-unknown-linux-gnu/miniupnpc/2.0-792e2b4396f/./.stamp_built] Error 2
```

Cloning [miniupnpc](https://github.com/miniupnp) and removing the `-D_XOPEN_SOURCE=600` clears the issue:

```
git clone https://github.com/miniupnp/miniupnp.git
cd miniupnp/
find $PWD -name 'minisoap.c'
   miniupnp/miniupnpc/minisoap.c
cd miniupnp/
gcc -m64 -fPIC -pipe -O1 -I../../depends/work/build/x86_64-unknown-linux-gnu/min
iupnpc/2.0-792e2b4396f -I/export/home/zcash/depends/x86_64-unknown-linux-gnu/inc
lude -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DM
INIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE -c -o minisoap.o minisoap.
c
```

-----

Also see [MiniUPnP Issue 216: Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications](https://github.com/miniupnp/miniupnp/issues/216).

```
$ git diff > miniupnpc.diff

$ cat miniupnpc.diff
diff --git a/miniupnpc/Makefile b/miniupnpc/Makefile
index 5c23000..a520dff 100644
--- a/miniupnpc/Makefile
+++ b/miniupnpc/Makefile
@@ -13,6 +13,12 @@
 OS = $(shell uname -s)
 VERSION = $(shell cat VERSION)
 
+ifeq ($(OS), SunOS)
+ifeq ($(CC), cc)
+CC=gcc
+endif
+endif
+
 ifeq ($(OS), Darwin)
 JARSUFFIX=mac
 LIBTOOL ?= $(shell which libtool)
@@ -43,10 +49,12 @@ CFLAGS += -D_NETBSD_SOURCE
 endif
 ifneq ($(OS), FreeBSD)
 ifneq ($(OS), Darwin)
+ifneq ($(OS), SunOS)
 #CFLAGS += -D_POSIX_C_SOURCE=200112L
 CFLAGS += -D_XOPEN_SOURCE=600
 endif
 endif
+endif
 #CFLAGS += -ansi
 # -DNO_GETADDRINFO
 INSTALL = install
diff --git a/miniupnpc/minissdpc.c b/miniupnpc/minissdpc.c
index f200f07..263160e 100644
--- a/miniupnpc/minissdpc.c
+++ b/miniupnpc/minissdpc.c
@@ -73,6 +73,9 @@ struct sockaddr_un {
 
 #if !defined(HAS_IP_MREQN) && !defined(_WIN32)
 #include <sys/ioctl.h>
+#if defined(__sun)
+#include <sys/sockio.h>
+#endif
 #endif
 
 #if defined(HAS_IP_MREQN) && defined(NEED_STRUCT_IP_MREQN)
```

-----

Here's what the test looks like. It builds a `git clone` of  miniUPnP.

```
$ gmake clean &>/dev/null && gmake
/bin/sh updateminiupnpcstrings.sh
Detected OS [SunOS] version [5.11]
MiniUPnPc version [2.0]
setting OS_STRING macro value to SunOS/5.11 in miniupnpcstrings.h.
setting MINIUPNPC_VERSION_STRING macro value to 2.0 in miniupnpcstrings.h.
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o miniwget.o miniwget.c
miniwget.c: In function ‘getHTTPResponse’:
miniwget.c:87:6: warning: variable ‘reason_phrase_len’ set but not used [-Wunused-but-set-variable]
  int reason_phrase_len = 0;
      ^
miniwget.c:86:9: warning: variable ‘reason_phrase’ set but not used [-Wunused-but-set-variable]
  char * reason_phrase = NULL;
         ^
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o minixml.o minixml.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o igd_desc_parse.o igd_desc_parse.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o minisoap.o minisoap.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o miniupnpc.o miniupnpc.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o upnpreplyparse.o upnpreplyparse.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o upnpcommands.o upnpcommands.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o upnperrors.o upnperrors.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o connecthostport.o connecthostport.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o portlistingparse.o portlistingparse.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o receivedata.o receivedata.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o upnpdev.o upnpdev.c
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o minissdpc.o minissdpc.c
ar crs libminiupnpc.a miniwget.o minixml.o igd_desc_parse.o minisoap.o miniupnpc.o upnpreplyparse.o upnpcommands.o upnperrors.o connecthostport.o portlistingparse.o receivedata.o upnpdev.o minissdpc.o
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o upnpc.o upnpc.c
gcc -lsocket -lnsl -lresolv -o upnpc-static upnpc.o libminiupnpc.a  
gcc -fPIC -O -Wall -W -Wstrict-prototypes -fno-common -DMINIUPNPC_SET_SOCKET_TIMEOUT -DMINIUPNPC_GET_SRC_ADDR -D_BSD_SOURCE -D_DEFAULT_SOURCE   -c -o listdevices.o listdevices.c
gcc -lsocket -lnsl -lresolv  listdevices.o libminiupnpc.a   -o listdevices
gcc -shared -lsocket -lnsl -lresolv -Wl,-soname,libminiupnpc.so.16 -o libminiupnpc.so miniwget.o minixml.o igd_desc_parse.o minisoap.o miniupnpc.o upnpreplyparse.o upnpcommands.o upnperrors.o connecthostport.o portlistingparse.o receivedata.o upnpdev.o minissdpc.o
gcc -lsocket -lnsl -lresolv -o upnpc-shared upnpc.o libminiupnpc.so  
```

--- a/depends/packages/miniupnpc.mk
+++ b/depends/packages/miniupnpc.mk
@@ -3,6 +3,7 @@ $(package)_version=2.0
 $(package)_download_path=http://miniupnp.free.fr/files
 $(package)_file_name=$(package)-$($(package)_version).tar.gz
 $(package)_sha256_hash=d434ceb8986efbe199c5ca53f90ed53eab290b1e6d0530b717eb6fa49d61f93b
+$(package)_patches=fix-solaris-compilation.patch strlen-before-memcmp.patch
 
 define $(package)_set_vars
 $(package)_build_opts=CC="$($(package)_cc)"
@@ -14,7 +15,9 @@ endef
 define $(package)_preprocess_cmds
   mkdir dll && \
   sed -e 's|MINIUPNPC_VERSION_STRING \"version\"|MINIUPNPC_VERSION_STRING \"$($(package)_version)\"|' -e 's|OS/version|$(host)|' miniupnpcstrings.h.in > miniupnpcstrings.h && \
-  sed -i.old "s|miniupnpcstrings.h: miniupnpcstrings.h.in wingenminiupnpcstrings|miniupnpcstrings.h: miniupnpcstrings.h.in|" Makefile.mingw
+  sed -i.old "s|miniupnpcstrings.h: miniupnpcstrings.h.in wingenminiupnpcstrings|miniupnpcstrings.h: miniupnpcstrings.h.in|" Makefile.mingw && \
+  patch -p2 < $($(package)_patch_dir)/fix-solaris-compilation.patch && \
+  patch -p2 < $($(package)_patch_dir)/strlen-before-memcmp.patch
 endef
 
 define $(package)_build_cmds

--- a/depends/packages/miniupnpc.mk
+++ b/depends/packages/miniupnpc.mk
@@ -3,7 +3,7 @@ $(package)_version=2.0
 $(package)_download_path=http://miniupnp.free.fr/files
 $(package)_file_name=$(package)-$($(package)_version).tar.gz
 $(package)_sha256_hash=d434ceb8986efbe199c5ca53f90ed53eab290b1e6d0530b717eb6fa49d61f93b
-$(package)_patches=fix-solaris-compilation.patch strlen-before-memcmp.patch
+$(package)_patches=fix-solaris-compilation.patch strlen-before-memcmp.patch patch-strlen-patch.patch
 
 define $(package)_set_vars
 $(package)_build_opts=CC="$($(package)_cc)"
@@ -17,7 +17,8 @@ define $(package)_preprocess_cmds
   sed -e 's|MINIUPNPC_VERSION_STRING \"version\"|MINIUPNPC_VERSION_STRING \"$($(package)_version)\"|' -e 's|OS/version|$(host)|' miniupnpcstrings.h.in > miniupnpcstrings.h && \
   sed -i.old "s|miniupnpcstrings.h: miniupnpcstrings.h.in wingenminiupnpcstrings|miniupnpcstrings.h: miniupnpcstrings.h.in|" Makefile.mingw && \
   patch -p2 < $($(package)_patch_dir)/fix-solaris-compilation.patch && \
-  patch -p2 < $($(package)_patch_dir)/strlen-before-memcmp.patch
+  patch -p2 < $($(package)_patch_dir)/strlen-before-memcmp.patch && \
+  patch -p2 < $($(package)_patch_dir)/patch-strlen-patch.patch
 endef
 
 define $(package)_build_cmds

--- a/depends/patches/miniupnpc/fix-solaris-compilation.patch
+++ b/depends/patches/miniupnpc/fix-solaris-compilation.patch
@@ -1,0 +1,42 @@
+From 71ce1d6dfa5424f8fe8633e23494c7638ea2c79e Mon Sep 17 00:00:00 2001
+From: Thomas Bernard <miniupnp@free.fr>
+Date: Thu, 10 Nov 2016 21:55:33 +0100
+Subject: [PATCH] fix for Solaris 11 compilation
+
+see #216
+---
+ miniupnpc/Makefile    | 2 ++
+ miniupnpc/minissdpc.c | 3 +++
+ 2 files changed, 5 insertions(+)
+
+diff --git a/miniupnpc/Makefile b/miniupnpc/Makefile
+index 5c23000..72cdc0f 100644
+--- a/miniupnpc/Makefile
++++ b/miniupnpc/Makefile
+@@ -43,10 +43,12 @@ CFLAGS += -D_NETBSD_SOURCE
+ endif
+ ifneq ($(OS), FreeBSD)
+ ifneq ($(OS), Darwin)
++ifneq ($(OS), SunOS)
+ #CFLAGS += -D_POSIX_C_SOURCE=200112L
+ CFLAGS += -D_XOPEN_SOURCE=600
+ endif
+ endif
++endif
+ #CFLAGS += -ansi
+ # -DNO_GETADDRINFO
+ INSTALL = install
+diff --git a/miniupnpc/minissdpc.c b/miniupnpc/minissdpc.c
+index f200f07..263160e 100644
+--- a/miniupnpc/minissdpc.c
++++ b/miniupnpc/minissdpc.c
+@@ -73,6 +73,9 @@ struct sockaddr_un {
+ 
+ #if !defined(HAS_IP_MREQN) && !defined(_WIN32)
+ #include <sys/ioctl.h>
++#if defined(__sun)
++#include <sys/sockio.h>
++#endif
+ #endif
+ 
+ #if defined(HAS_IP_MREQN) && defined(NEED_STRUCT_IP_MREQN)

--- a/depends/patches/miniupnpc/patch-strlen-patch.patch
+++ b/depends/patches/miniupnpc/patch-strlen-patch.patch
@@ -1,0 +1,22 @@
+From 0aa7c46227acd8ddb135c577674ad454bf2fba86 Mon Sep 17 00:00:00 2001
+From: Thomas Bernard <miniupnp@free.fr>
+Date: Fri, 11 Nov 2016 17:53:21 +0100
+Subject: [PATCH] remove unsigned/signed comparison
+
+---
+ miniupnpc/portlistingparse.c | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/miniupnpc/portlistingparse.c b/miniupnpc/portlistingparse.c
+index 1bed763..07f3f87 100644
+--- a/miniupnpc/portlistingparse.c
++++ b/miniupnpc/portlistingparse.c
+@@ -55,7 +55,7 @@ startelt(void * d, const char * name, int l)
+ 	pdata->curelt = PortMappingEltNone;
+ 	for(i = 0; elements[i].str; i++)
+ 	{
+-		if(strlen(elements[i].str) == l && memcmp(name, elements[i].str, l) == 0)
++		if(strlen(elements[i].str) == (size_t)l && memcmp(name, elements[i].str, l) == 0)
+ 		{
+ 			pdata->curelt = elements[i].code;
+ 			break;

--- a/depends/patches/miniupnpc/strlen-before-memcmp.patch
+++ b/depends/patches/miniupnpc/strlen-before-memcmp.patch
@@ -1,0 +1,23 @@
+From ec1c49bb0cd5e448e6f0adee7de3a831c4869bdd Mon Sep 17 00:00:00 2001
+From: Thomas Bernard <miniupnp@free.fr>
+Date: Fri, 11 Nov 2016 17:24:39 +0100
+Subject: [PATCH] check strlen before memcmp
+
+1st try to fix #220
+---
+ miniupnpc/portlistingparse.c | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/miniupnpc/portlistingparse.c b/miniupnpc/portlistingparse.c
+index 0e09278..1bed763 100644
+--- a/miniupnpc/portlistingparse.c
++++ b/miniupnpc/portlistingparse.c
+@@ -55,7 +55,7 @@ startelt(void * d, const char * name, int l)
+ 	pdata->curelt = PortMappingEltNone;
+ 	for(i = 0; elements[i].str; i++)
+ 	{
+-		if(memcmp(name, elements[i].str, l) == 0)
++		if(strlen(elements[i].str) == l && memcmp(name, elements[i].str, l) == 0)
+ 		{
+ 			pdata->curelt = elements[i].code;
+ 			break;

